### ISSUE TITLE
Bootstrapping a function with 2 inputs?

### ISSUE THREAD

**Issue by drbenvincent** [juliangehring/Bootstrap.jl#70]
Is the package currently able to do bootstrap with functions that take 2 inputs? A concrete example would be getting bootstrap intervals of a correlation coefficient based on x, y data. 

I tried to give it a go with how I imagined it would work (below). It 'works' in that it doesn't give an error, but I'm unable to interpret the output and figure out if it has done anything meaningful.

```julia
using Distributions
using Statistics
using Bootstrap

N = 100
personality = rand(Normal(), N)
looks = rand(Normal(), N)

bs = bootstrap(cor, [personality  looks], BasicSampling(1000))
bci = confint(bs, BasicConfInt(0.95))
```

`bs` gives...
```
Bootstrap Sampling
  Estimates:
    │ Var │ Estimate  │ Bias        │ StdError  │
    │     │ Float64   │ Float64     │ Float64   │
    ├─────┼───────────┼─────────────┼───────────┤
    │ 1   │ 1.0       │ 0.0         │ 0.0       │
    │ 2   │ 0.0308112 │ -0.00386273 │ 0.0986716 │
    │ 3   │ 0.0308112 │ -0.00386273 │ 0.0986716 │
    │ 4   │ 1.0       │ 0.0         │ 0.0       │
  Sampling: BasicSampling
  Samples:  1000
  Data:     Array{Float64,2}: { 100 × 2 }
```

and `bci` is

    ((1.0, 1.0, 1.0), (0.030811242352543605, -0.15276601303154966, 0.2214372656540179), (0.030811242352543605, -0.15276601303154966, 0.2214372656540179), (1.0, 1.0, 1.0))

**Comment by juliangehring** [comment]
I assume that you would want to keep the original pairing between `x` and `y` in your use case? I.e. if you pick `x[i]` you also want to pick `y[i]` in a sampling step?

**Comment by juliangehring** [comment]
If the data is an `Array`, the implementation will treat the rows as observations and the columns as variables; analogous to a `DataFrame`. The `bootstrap` step will sample different observations while keeping the relation between the variables intact. In other words, it will generate a new data set with `data[idx,:]` for each sampling.

**Comment by juliangehring** [comment]
In your case, the package should work as you seem to expect. Let me know if my explanations make sense or if you anything isn't clear.

**Comment by drbenvincent** [comment]
Thanks for the quick reply.

> I assume that you would want to keep the original pairing between `x` and `y` in your use case? I.e. if you pick `x[i]` you also want to pick `y[i]` in a sampling step?

Yes

> If the data is an Array, the implementation will treat the rows as observations and the columns as variables; analogous to a DataFrame. The bootstrap step will sample different observations while keeping the relation between the variables intact. In other words, it will generate a new data set with data[idx,:] for each sampling.

Sounds like this is doing what I wanted it to

> In your case, the package should work as you seem to expect. Let me know if my explanations make sense or if you anything isn't clear.

I guess I am a bit confused about why there were 4 output variables rather than just 1 for the bootstrap correlation values. In this case the variables are uncorrelated, what I was after overall was the 95% CI's on the correlation coefficient. So am expecting some range overlapping zero (variables are independent/uncorrelated in this example), but was unsure how to interpret the output.

**Comment by juliangehring** [comment]
> I guess I am a bit confused about why there were 4 output variables rather than just 1 for the bootstrap correlation values. In this case the variables are uncorrelated, what I was after overall was the 95% CI's on the correlation coefficient. So am expecting some range overlapping zero (variables are independent/uncorrelated in this example), but was unsure how to interpret the output.

The `Statistics.cor` function computes a 2x2 correlation matrix for your input array:

```julia
2×2 Array{Float64,2}:
 1.0       0.192743
 0.192743  1.0
```

Please note that this is a full correlation matrix and not only a coefficient between both variables.

This is why the output of the `bootstrap` step also gives you 4 variables, it is simply the output of `cor` applied on the bootstrapped data set. You are probably only interested in the off-diagonal elements, and could use e.g. `x -> cor(x)[2]` or `x -> cor(x[1,:], x[2,:])` as your statistics function.

The confidence intervals return
```julia
(estimate, lower_bound, upper_bound)
```
for each of the 4 bootstrapped variables. In line with what I mentioned above, you would want the 2nd or 3rd element in this case.

**Comment by drbenvincent** [comment]
Perfect. I got confused about the output of `Statistics.cor` because when I was testing I used it in the form of `cor(personality, looks)` and just got the scaler output. 

So this makes sense, but I think `x -> cor(x[1,:], x[2,:])` should be `x -> cor(x[:,1], x[:,2])`. That gives me:

    ((0.005388735203249658, -0.22214553137849985, 0.21166247504417582),)

and those numbers make sense for 100 uncorrelated values. Not sure why the extra empty element though. But that solves my question. Thanks very much!
